File: my-folder/1186-building-h2o/solution.java
import java.util.concurrent.Semaphore;
class H2O {
    private Semaphore hyd_sema;
    private Semaphore oxy_sema;
    private int count;
    private final Object lock = new Object();

    public H2O() {
        this.hyd_sema = new Semaphore(1);
        this.oxy_sema = new Semaphore(0);
        this.count = 0;
        
    }

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
		
        // releaseHydrogen.run() outputs "H". Do not change or remove this line.
        this.hyd_sema.acquire();
        releaseHydrogen.run();
        synchronized(lock){ 
            this.count += 1;
            if (this.count == 2){
                this.oxy_sema.release();
                this.count = 0;
            } else{
                this.hyd_sema.release();
            }
        }
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        
        // releaseOxygen.run() outputs "O". Do not change or remove this line.
        this.oxy_sema.acquire();
		releaseOxygen.run();
        this.hyd_sema.release();
    }
}

File: my-folder/1186-building-h2o/solution.py
"""goal: group them to form H2O
barrier: each thread has to wait until complete molecule can be formed.
use semaphore 
allow 2 hydrogen - release lock
allow 1 oxygen - release
"""
import threading
class H2O:
    def __init__(self):
        self.hyd_sema = threading.Semaphore(1)
        self.oxy_sema = threading.Semaphore(1)
        self.count = 0

    def hydrogen(self, releaseHydrogen: 'Callable[[], None]') -> None:
        
        # releaseHydrogen() outputs "H". Do not change or remove this line.
        self.hyd_sema.acquire()
        releaseHydrogen()
        self.count += 1
        if self.count == 2:
            self.oxy_sema.release()
            self.count = 0
        else:
            self.hyd_sema.release()


    def oxygen(self, releaseOxygen: 'Callable[[], None]') -> None:
        
        # releaseOxygen() outputs "O". Do not change or remove this line.
        self.oxy_sema.acquire()
        releaseOxygen()
        self.hyd_sema.release()
        
